### ISSUE TITLE
Dataloader shuffle is not reproducible. 

### ISSUE THREAD

**Issue by yorard123** [rusty1s/pytorch_geometric#294]
## 🐛 Bug

Dataloader shuffle is not reproducible. As far as I know dataloader in pytorch is reproducible if you set the seed. However, in pytorch geometric in each start the results are different using the same seed. 

## To Reproduce

In dataloader put the argument shuffle to true

## Environment

* OS: Ubuntu 18
* Python version: 3.6
* PyTorch version: 1.1
* CUDA/cuDNN version: 10
* GCC version: 7.3
* Any other relevant information:

**Comment by rusty1s** [comment]
Actually, this is not reproducible for me, and I do not see a reason what should prevent this from happening (as we simply use a different `collate` function and call the internal PyTorch `DataLoader`), *e.g.*:

```python
torch.manual_seed(12345)

loader = DataLoader(dataset, batch_size=1, shuffle=True)
for data in loader:
    print(data)
```

**Comment by yorard123** [comment]
I tried your example and it is working.. using modelnet dataset. I will double check my dataset, maybe I am doing something wrong... 

Thanks and sorry for the inconvenience.

**Comment by yorard123** [comment]
I checked again, and the dataloader and dataset are not the problem... I found that they are generating the same results in each execution.. However, in each execution I am obtaining different results... What are the random libraries used in pytorch_geometric? At this moment I am seeding numpy, torch, torch.cuda and random libraries. Am I missing something?

**Comment by rusty1s** [comment]
See [here](https://github.com/rusty1s/pytorch_geometric/issues/92) :(

**Comment by yorard123** [comment]
Oh I see! I always forget about this behaviour on gpu. However I do not understand the reason of that. In my case, I am using the NNConv, the operations done for this conv are mult and sums in essence. Why is it affecting the execution order to the results?  In addition, if I am using the the same gpu ... the floating point precision should be same, right? Then I do not understand why is it happening this  :( I suppose I am not taking into account something.

Thanks for your time!!

**Comment by rusty1s** [comment]
The problem is the aggregation, because the order of neighbors is not determined on GPU. Hence features can be summed up in different oder, even for the same GPU.

**Comment by yorard123** [comment]
Oh, you mean that the updated feature of one node is used in some aggregations and others the old state is used, right?

**Comment by rusty1s** [comment]
No, I mean that for each node, we aggregate neighboring node features in a *random* order on the GPU., which may result in small numerical instabilities.
